daemon/internal/newrelic/php_packages.go

```diff
@@ -34,51 +37,112 @@ func (packages *PhpPackages) NumSaved() float64 {
 // newPhpPackages returns a new PhpPackages struct.
 func NewPhpPackages() *PhpPackages {
 	p := &PhpPackages{
-		numSeen: 0,
-		data:    nil,
+		numSeen:      0,
+		data:         nil,
+		filteredPkgs: nil,
 	}
 
 	return p
 }
 
-// SetPhpPackages sets the observed package list.
-func (packages *PhpPackages) SetPhpPackages(data []byte) error {
+// filter seen php packages data to avoid sending duplicates
+//
+// the `App` structure contains a map of PHP Packages the reporting
+// application has encountered.
+//
+// the map of packages should persist for the duration of the
+// current connection
+//
+// the JSON format received from the agent is:
+//
+//	[["package_name","version",{}],...]
+//
+// for each entry, assign the package name and version to the `PhpPackagesKey`
+// struct and use the key to verify data does not exist in the map. If the
+// key does not exist, add it to the map and the slice of filteredPkgs to be
+// sent in the current harvest.
+func (packages *PhpPackages) Filter(pkgHistory map[PhpPackagesKey]struct{}) {
+	if packages == nil || packages.data == nil {
+		return
+	}
 
-	if nil == packages {
-		return fmt.Errorf("packages is nil!")
+	for _, pkgKey := range packages.data {
+		_, ok := pkgHistory[pkgKey]
+		if !ok {
+			pkgHistory[pkgKey] = struct{}{}
+			packages.filteredPkgs = append(packages.filteredPkgs, pkgKey)
+		}
 	}
-	if nil != packages.data {
-		log.Debugf("SetPhpPackages - data field was not nil |^%s| - overwriting data", packages.data)
+}
+
+// AddPhpPackagesFromData observes the PHP packages info from the agent.
+func (packages *PhpPackages) AddPhpPackagesFromData(data []byte) error {
+	if packages == nil {
+		return fmt.Errorf("packages is nil!")
 	}
-	if nil == data {
+	if data == nil || !(len(data) > 0) {
 		return fmt.Errorf("data is nil!")
 	}
+
+	var x []interface{}
+
+	err := json.Unmarshal(data, &x)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal php package json: %s", err.Error())
+	}
+
+	for _, pkgJson := range x {
+		pkg, _ := pkgJson.([]interface{})
+		if len(pkg) != 3 {
+			return fmt.Errorf("invalid php package json structure: %+v", pkg)
+		}
+
+		name, ok := pkg[0].(string)
+		if !ok || len(name) == 0 {
+			return fmt.Errorf("unable to parse package name")
+		}
+
+		version, ok := pkg[1].(string)
+		if !ok || len(version) == 0 {
+			return fmt.Errorf("unable to parse package version")
+		}
+
+		packages.data = append(packages.data, PhpPackagesKey{name, version})
+	}
+
 	packages.numSeen = 1
```

Comment:
I'm not sure what this field does any more - but if we keep it should it be incremented by the number of packages created in the `for` loop above?

Reply:
I'm also interested in how this gets resolved. See related comment: https://github.com/newrelic/newrelic-php-agent/pull/1103/files#r2261468915